mempool/cat/pool.go

```diff
@@ -0,0 +1,765 @@
+		var evictedBytes int64
+		for _, tx := range victims {
+			txmp.evictTx(tx)
+
+			// We may not need to evict all the eligible transactions.  Bail out
+			// early if we have made enough room.
+			evictedBytes += tx.size()
+			if evictedBytes >= wtx.size() {
+				break
+			}
+		}
+	}
+
+	txmp.store.set(wtx)
+
+	txmp.metrics.TxSizeBytes.Observe(float64(wtx.size()))
+	txmp.metrics.Size.Set(float64(txmp.Size()))
+	txmp.logger.Debug(
+		"inserted new valid transaction",
+		"priority", wtx.priority,
+		"tx", fmt.Sprintf("%X", wtx.key),
+		"height", wtx.height,
+		"num_txs", txmp.Size(),
+	)
+	txmp.notifyTxsAvailable()
+	return nil
+}
+
+func (txmp *TxPool) evictTx(wtx *wrappedTx) {
+	txmp.store.remove(wtx.key)
+	txmp.metrics.EvictedTxs.Add(1)
+	txmp.evictedTxs.Push(wtx)
+	txmp.logger.Debug(
+		"evicted valid existing transaction; mempool full",
+		"old_tx", fmt.Sprintf("%X", wtx.key),
+		"old_priority", wtx.priority,
+	)
+}
+
+// handleRecheckResult handles the responses from ABCI CheckTx calls issued
+// during the recheck phase of a block Update.  It removes any transactions
+// invalidated by the application.
+//
+// This method is NOT executed for the initial CheckTx on a new transaction;
+// that case is handled by addNewTransaction instead.
+func (txmp *TxPool) handleRecheckResult(wtx *wrappedTx, checkTxRes *abci.ResponseCheckTx) {
+	txmp.metrics.RecheckTimes.Add(1)
+
+	// If a postcheck hook is defined, call it before checking the result.
+	err := txmp.postCheck(wtx.tx, checkTxRes)
+
+	if checkTxRes.Code == abci.CodeTypeOK && err == nil {
+		// Note that we do not update the transaction with any of the values returned in
+		// recheck tx
+		return // N.B. Size of mempool did not change
+	}
+
+	txmp.logger.Debug(
+		"existing transaction no longer valid; failed re-CheckTx callback",
+		"priority", wtx.priority,
+		"tx", fmt.Sprintf("%X", wtx.key),
+		"err", err,
+		"code", checkTxRes.Code,
+	)
+	txmp.store.remove(wtx.key)
+	txmp.rejectedTxCache.Push(wtx.key)
+	txmp.metrics.FailedTxs.Add(1)
+	txmp.metrics.Size.Set(float64(txmp.Size()))
+}
+
+// recheckTransactions initiates re-CheckTx ABCI calls for all the transactions
+// currently in the mempool. It reports the number of recheck calls that were
+// successfully initiated.
+//
+// Precondition: The mempool is not empty.
+// The caller must hold txmp.mtx exclusively.
+func (txmp *TxPool) recheckTransactions() {
+	if txmp.Size() == 0 {
+		panic("mempool: cannot run recheck on an empty mempool")
+	}
+	txmp.logger.Debug(
+		"executing re-CheckTx for all remaining transactions",
+		"num_txs", txmp.Size(),
+		"height", txmp.Height(),
+	)
+
+	// Collect transactions currently in the mempool requiring recheck.
+	wtxs := txmp.store.getAllTxs()
+
+	// Issue CheckTx calls for each remaining transaction, and when all the
+	// rechecks are complete signal watchers that transactions may be available.
+	go func() {
+		g, start := taskgroup.New(nil).Limit(2 * runtime.NumCPU())
```

Comment:
why twice the number of recognized CPUs?

Reply:
I have no idea. 

I think the problem with making a direct copy of the priority mempool is that there are a lot of changes here that are just inherited from previous authors. Not that we shouldn't go over the work in it's entirety it's just that I've made it difficult to identify the changed that I've actually made and the ones that previously existed. I think what I would do next time is to have the first PR just be directly copying the v1 mempool as cat and then build the rest of the PR's of that. I guess I could still do that if people wanted. The nice thing about having the complete change is people can learn more about the workings of the mempool in its entirety.